Introduce a Micronaut module descriptor

melix opened the pull request:
This commit introduces a facility to list the Micronaut modules which are found at runtime. This will be useful diagnostics information, which can for example be used at runtime by the Micronaut Control Panel.

It is implemented 2-fold. First of all, an independent module named `micronaut-module-info`, standalone without any dependencies, defines the base interface and abstract class that can be used for such descriptors. The goal is to load the descriptors using service loading, independently of the framework. The reason we do this is because this module can then be used by the build plugins to either:
   - generate an implementation of a descriptor, which will then be loaded by Micronaut
   - statically reason about modules, for example for linting at build time.

The second module, `micronaut-module-info-runtime`, maps these services into Micronaut beans. A singleton, `MicronautRuntimeModules` can be used to fetch information about modules, which are then resolved as a hierarchy of modules.

For now, the implementation is limited, provides minimal information such a name for the module, its version, the Maven coordinates of the module and some tags. Eventually, we may enrich this with more information such as the dependencies of the module, or compatibility with other modules. For example this could become possible to express runtime requirements such as "module A requires at least one implementation of B at runtime, but not more than 1".

Or, it could become possible for a module to explain which test resources modules it requires, or even supply a list of MCP servers that could be used for coding with AI...

melix committed:
Introduce a Micronaut module descriptor

This commit introduces a facility to list the Micronaut modules
which are found at runtime. This will be useful diagnostics
information, which can for example be used at runtime by the
Micronaut Control Panel.

It is implemented 2-fold. First of all, an independent module
named `micronaut-module-info`, standalone without any dependencies,
defines the base interface and abstract class that can be used
for such descriptors. The goal is to load the descriptors using
service loading, independently of the framework. The reason we
do this is because this module can then be used by the build
plugins to either:
   - generate an implementation of a descriptor, which will
then be loaded by Micronaut
   - statically reason about modules, for example for linting
at build time.

The second module, `micronaut-module-info-runtime`, maps these
services into Micronaut beans. A singleton, `MicronautRuntimeModules`
can be used to fetch information about modules, which are
then resolved as a hierarchy of modules.

For now, the implementation is limited, provides minimal information
such a name for the module, its version, the Maven coordinates of
the module and some tags. Eventually, we may enrich this with more
information such as the dependencies of the module, or compatibility
with other modules. For example this could become possible to express
runtime requirements such as "module A requires at least one implementation
of B at runtime, but not more than 1".

Or, it could become possible for a module to explain which test resources
modules it requires, or even supply a list of MCP servers that could be
used for coding with AI...

melix committed:
Disable ossIndexAudit on module-info module

graemerocher commented on module-info-runtime/src/main/java/io/micronaut/module/info/runtime/MicronautModuleRuntimeInfoFactory.java:
```diff
@@ -0,0 +1,124 @@
+ */
+@Internal
+@Factory
+class MicronautModuleRuntimeInfoFactory {
```
can be `final`

graemerocher commented on module-info-runtime/src/main/java/io/micronaut/module/info/runtime/MicronautModuleRuntimeInfoFactory.java:
```diff
@@ -0,0 +1,124 @@
+class MicronautModuleRuntimeInfoFactory {
+
+    @Singleton
+    public List<MicronautRuntimeModule> getAllModules(ApplicationContext context) {
```
you could just define `List<MicronautModuleInfo> moduleInfos` as a method argument

graemerocher commented on module-info-runtime/src/main/java/io/micronaut/module/info/runtime/MicronautModuleRuntimeInfoFactory.java:
```diff
@@ -0,0 +1,124 @@
+
+    @Singleton
+    public List<MicronautRuntimeModule> getAllModules(ApplicationContext context) {
+        var moduleInfos = context.getBeansOfType(MicronautModuleInfo.class);
```
we might want to a GraalVM optimisation to load this statically at build time

melix replied:
There is an optimization in Micronaut AOT for service loading, which proves to be redundant nowadays.

graemerocher commented on module-info-runtime/src/main/java/io/micronaut/module/info/runtime/MicronautRuntimeModule.java:
```diff
@@ -0,0 +1,130 @@
+ */
+@Experimental
+@NonNull
+public final class MicronautRuntimeModule {
```
add nullability annotations to API contract

melix replied:
What do you mean? `@NonNull` isn't enough?

alvarosanchez replied:
I didn't know you could use `@NonNull` at a type level. What's the behaviour then?

It's own Javadoc says:

> Should be used at parameter, return value, and field level

alvarosanchez replied:
What I wonder, on the other hand, is? can this be a record?

melix replied:
mmm, I thought this worked like in JSpecify... @graemerocher what do you think?

graemerocher commented on module-info-runtime/src/main/java/io/micronaut/module/info/runtime/MicronautRuntimeModule.java:
```diff
@@ -0,0 +1,130 @@
+        return info.getTags();
+    }
+
+    public MicronautRuntimeModule getRoot() {
```
javadoc

graemerocher commented on module-info/src/main/java/io/micronaut/module/info/MicronautModuleInfo.java:
```diff
@@ -0,0 +1,75 @@
+ * This can be used for diagnostics. Information is loaded via service loading.
+ * Each module is recommended to extend the abstract class instead of directly implementing this interface.
+ */
+public interface MicronautModuleInfo {
```
add nullability annotations to API contract. Add `@Experimental`

melix replied:
We can't here: this module intentionally doesn't depend on anything Micronaut related.

graemerocher reviewed (changes requested):
How is it anticipated the module info will be populated? By some build tool?

graemerocher commented:
I management endpoint would be useful I think to see what modules are active and what beans each provide.

melix commented:
> How is it anticipated the module info will be populated? By some build tool?

I am working on a Micronaut Build change to automatically generate classes for each module.

melix committed:
Address review feedback

alvarosanchez commented on module-info-runtime/src/test/java/io/micronaut/module/info/runtime/ServiceLoaderDefinedModule.java:
```diff
@@ -0,0 +1,33 @@
+
+import java.util.Set;
+
+public class ServiceLoaderDefinedModule extends AbstractMicronautModuleInfo {
```
This is more or less what you will generate with micronaut-build right?

melix replied:
correct

melix replied:
It cannot be a record unless you want to give away the `Optional` methods (null safety).

alvarosanchez commented on module-info/src/main/java/io/micronaut/module/info/AbstractMicronautModuleInfo.java:
```diff
@@ -0,0 +1,98 @@
+ * Base class for Micronaut module information.
+ * It is recommended, though not required, to extend this class.
+ */
+public abstract class AbstractMicronautModuleInfo implements MicronautModuleInfo {
```
Can be a record?

melix replied:
No because each generated class will have to be different and a record cannot be extended. In addition module descriptors may add some logic later (e.g compatibility with other modules).

sdelamo committed:
test: isolate ConstantPropertySourceSpec in its own module (#12195)

test ConstantPropertySourceSpec passes when run on isolation. It fails when run in a suite. This PR moves the test to its own module. Not ideal but at least we remove the flakiness of the test:

https://ge.micronaut.io/scans/tests?search.relativeStartTime=P90D&search.timeZoneId=Europe%2FMadrid&tests.container=io.micronaut.context.env.ConstantPropertySourceSpec&tests.help=tags

sdelamo committed:
test: ProxyBackpressureTest run localy reduced cvssource (#12196)

Tests with:

```
        "true,2,/large",
        "true,2,/proxy",
``

Fail locally:

https://ge.micronaut.io/scans/tests?search.timeZoneId=Europe%2FMadrid&tests.container=io.micronaut.http.client.ProxyBackpressureTest&tests.help=tags

While they are flaky in CI:

https://ge.micronaut.io/scans/tests?search.tags=CI&search.timeZoneId=Europe%2FMadrid&tests.container=io.micronaut.http.client.ProxyBackpressureTest&tests.help=tags

This pull-request modifies the test so that locally only a reduce CSV source is run

sdelamo committed:
test: ignore LoomCarrierSpec for MacOS (#12197)

* test: ignore LoomCarrierSpec for MacOS

LoomCarrierSpec fails on MacOS passes for Linux

https://ge.micronaut.io/scans/tests?search.startTimeMax=1763506799999&search.startTimeMin=1762815600000&search.tags=not:CI&search.timeZoneId=Europe%2FMadrid&tests.container=io.micronaut.http.server.netty.LoomCarrierSpec&tests.test=sticky%20on%20poller%20thread&tests.unstableOnly=false

* ignore if Java 25

micronaut-build committed:
Update common files (#12178)

sdelamo committed:
build: disable htmlSanityCheck (#12198)

see: https://github.com/micronaut-projects/micronaut-core/issues/12181
